### PR TITLE
Upgrade to linkerd 2.10.2

### PR DESCRIPTION
**Pull request by ubergesundheit**
This PR updates the linkerd2-app to use upstream version `stable-2.10.2`

Additional changes:

- Add tests
- Publish to `giantswarm` catalog
- Adds basic sli monitoring labels

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,6 +13,8 @@ need to generate a CA certificate prior to installation.
 - `cert-manager` must already be deployed in the cluster.
 - it is strongly suggested to use the [`linkerd2-cni-app`](https://github.com/giantswarm/linkerd2-cni-app) as this results in a more secure setup.
   - the CNI must be installed before this chart, and `global.cniEnabled: true` must be set.
+- with linkerd CNI enabled all pods linked by linkerd require a PodSecurityPolicy, which allows use of `EmptyDir`
+  - the injected linkerd-proxy container requires an `EmptyDir`
```
Comment (ubergesundheit) [inline README.md:17]: @ced0ps 

```suggestion
- with linkerd CNI enabled, all pods meshed by linkerd require a `PodSecurityPolicy`, which allows use of `EmptyDir` volumes
  - the injected linkerd-proxy container requires an `EmptyDir` volumes
```

Comment (ced0ps) [inline README.md]:
```suggestion
- with linkerd CNI enabled, all pods meshed by linkerd require a `PodSecurityPolicy`, which allows use of `EmptyDir` volumes
  - the injected linkerd-proxy container requires an `EmptyDir` volume
```

File: README.md [hunk 2]
```diff
@@ -2,46 +2,61 @@
...
+This chart is based on the official linkerd2 helm chart.
 
 ## Requirements
 
 - you can install only one release of this chart per kubernetes cluster.
-- `cert-manager` must already be deployed in the cluster.
 - it is strongly suggested to use the [`linkerd2-cni-app`](https://github.com/giantswarm/linkerd2-cni-app) as this results in a more secure setup.
   - the CNI must be installed before this chart, and `global.cniEnabled: true` must be set.
+- with linkerd CNI enabled, all pods meshed by linkerd require a `PodSecurityPolicy`, which allows use of `EmptyDir` volumes
+  - the injected linkerd-proxy container requires an `EmptyDir` volume
```
Comment (ubergesundheit) [inline README.md:13]: sorry to be the party-pooper again, its two volumes.. one for the `proxy-init` initContainer and one for the `proxy` sidecar container

check https://github.com/giantswarm/linkerd2-app/blob/subtree-2.10/helm/partials/templates/_volumes.tpl

```suggestion
  - the injected linkerd-proxy container requires an `EmptyDir` volumes
```

File: README.md [hunk 3]
```diff
@@ -2,46 +2,61 @@
...
+This chart is based on the official linkerd2 helm chart.
 
 ## Requirements
 
 - you can install only one release of this chart per kubernetes cluster.
-- `cert-manager` must already be deployed in the cluster.
 - it is strongly suggested to use the [`linkerd2-cni-app`](https://github.com/giantswarm/linkerd2-cni-app) as this results in a more secure setup.
   - the CNI must be installed before this chart, and `global.cniEnabled: true` must be set.
+- with linkerd CNI enabled, all pods meshed by linkerd require a `PodSecurityPolicy`, which allows use of `EmptyDir` volumes
+  - the injected linkerd-proxy container requires an `EmptyDir` volumes
```
Comment (ubergesundheit) [inline README.md:13]: Ok I am a little doofus :crab:  @ced0ps please help :P

```suggestion
  - the injected linkerd-proxy container and proxy initContainer require an `EmptyDir` volume
```

**Comment by cokiengchiara** [comment]
YEAHHH exciting. Thanks for taking care of the docs as well. Individual instructions are clear. I have a suggestion on the structure - please check if it makes sense.

- Comment: It seems like the ‘do this, then do that’ steps are combined with “Troubleshooting / BTW, make sure that..." are combined together. If the goal is to get them started as fast as possible, I would recommend adjusting the lay out a bit, like the below, and separating the ‘troubleshooting’
- Question: Why would they want to install it without CNI plugin? Is it a rare case? If so, I would recommend creating a separate section… “Installing without the CNI plugin” This helps focus the quickstart only on the essentials.


---

### Quickstart Guide


**Step 1: Pre-installation**


1. Install linkerd2 cni. Make sure cniEnabled:true is set (default)
2. Add PSP to all workload pods, to allow use of emptydir volumes? 

**Step 2: Configure Linkerd**


3. Review default values.yaml file to see if you want to make any changes to the user configurations. Use DRY YAML anchors where possible.
4. Generate a trust anchor and issuer certificate
5. Construct config file

**Step 3: Deploy Linkerd**


6. Decide if you want to deploy via app cr or helm. We recommend app cr b/c…  
	1. For app cr… Set required annotations and labels in the chart
	2. For Helm… 
7. Add necessary annotations and labels to your namespace

**Step 4: After deployment**


8. Disable proxy injections for pods running in the kube-system namespace
9. optional but recommended: linkerd cli tool
10. Install linkerdviz extention
11. Open linkerd dashboard

### Installing without the CNI plugin


### Troubleshooting

- Only one release of this chart per cluster
- only possible namespace is linkerd

File: helm/linkerd2-app/values.yaml [hunk 4]
```diff
@@ -2,405 +2,321 @@
...
+# -- Log format for the control plane components
+controllerLogFormat: plain
+# -- enables control plane tracing
+controlPlaneTracing: false
+# -- namespace to send control plane traces to
+controlPlaneTracingNamespace: linkerd-jaeger
+# -- control plane version. See Proxy section for proxy version
+linkerdVersion: stable-2.10.2
+# -- Control plane namespace
+namespace: linkerd2-app
```
Comment (ced0ps) [inline helm/linkerd2-app/values.yaml:35]:
```suggestion
namespace: linkerd
```

**Review by ced0ps — APPROVED** [review]
I think it's pretty much where it needs to be right now.

File: README.md [hunk 5]
```diff
@@ -2,58 +2,146 @@
...
+
+After installation, you can open the dashboard by executing
+
+```
+linkerd viz dashboard
+```
+
+## Mesh your workloads
+
+After installation, linkerd looks for a `linkerd.io/inject: enabled` annotation on `Namespaces` or other workload resources. Adding this annotation to your workload namespaces will trigger automatic proxy container injection to your pods. More information on proxy injection can be found on the ["Automatic Proxy Injection" page](https://linkerd.io/2.10/features/proxy-injection/) in the upstream documentation.
```
Comment (piontec) [inline README.md:121]: Maybe let's add info about how to use `namespaceConfiguration` in App CRs to get the same effect for apps deployed with app-platform?

File: README.md [hunk 6]
```diff
@@ -2,58 +2,146 @@
...
+The `linkerd` namespace name is assumed in several parts of the chart, so we restricted installation to namespaces named `linkerd` only.
 
-### Note:
+### Proxy containers fail to start
 
-`values.yaml` uses [YAML anchors](https://helm.sh/docs/chart_template_guide/yaml_techniques/#yaml-anchors) to
-be DRY where possible, however this can cause problems when these values are overridden by a custom
-values file. If you wish to enabled or disable deployment of the `grafana` and `tracing` subcharts
-then you should override the values at `.$subchart.enabled`, **not** `.global.$subchart.enabled`.
+Your workload must be able to create `EmptyDir` volumes, so you'll need to create a `PodSecurityPolicy` allowing the creation of `EmptyDir` volumes.
```
Comment (piontec) [inline README.md:139]: We should add info here about how to edit the default PSP to easily allow for `emptyDirs` everywhere. Doing this one-by-one for every deployment is cumbersome.

File: README.md (hunk 5 above)
Comment (ubergesundheit) [reply README.md]: added :heavy_check_mark:

File: README.md (hunk 6 above)
Comment (ubergesundheit) [reply README.md:141]: added :heavy_check_mark:

File: README.md [hunk 7]
```diff
@@ -1,59 +1,175 @@
...
+kubectl gs template app \
+  --catalog giantswarm \
+  --name linkerd2-app \
+  --namespace linkerd \
+  --cluster <your-cluster-id>  \
+  --version 0.6.0 \
+  --user-configmap my-linkerd-values.yaml \
+  --user-secret my-linkerd-certificates.yaml \
+  --namespace-labels "linkerd.io/is-control-plane=true,config.linkerd.io/admission-webhooks=disabled,linkerd.io/control-plane-ns=linkerd" \
+  --namespace-annotations "linkerd.io/inject=disabled > linkerd-manifest.yaml"
```
Comment (ubergesundheit) [inline README.md:57]:
```suggestion
  --namespace-annotations "linkerd.io/inject=disabled" > linkerd-manifest.yaml
```

File: README.md [hunk 8]
```diff
@@ -1,59 +1,175 @@
...
+  version: 0.6.0
+```
 
-This chart is based on the official linkerd2 helm chart. The main difference it offers is a
-direct integration with [cert-manager](https://cert-manager.io/) used to generate short-lived
-certificates for linkerd2 components. This results in an entirely hands-off install - no
-need to generate a CA certificate prior to installation.
+Please note, that for security reasons Giant Swarm by default forbids the usage of
+`emptyDir` volumes as storage for pods. Linkerd needs this functionality to deploy
+linkerd itself, but also to run any deployment that is going to be included in
```
Comment (ubergesundheit) [inline README.md:94]:
```suggestion
`emptyDir` volumes as storage for pods. Linkerd needs this functionality to deploy the
linkerd control plane itself, but also to run the proxy containers in any deployment that is going to be included in
```

File: README.md [hunk 9]
```diff
@@ -1,59 +1,175 @@
...
+```
 
-## Requirements
+then finding the `spec.volumes` list and appending `emptyDir` value to the list. Then, write the file and exit your editor.
 
-- you can install only one release of this chart per kubernetes cluster.
-- `cert-manager` must already be deployed in the cluster.
-- it is strongly suggested to use the [`linkerd2-cni-app`](https://github.com/giantswarm/linkerd2-cni-app) as this results in a more secure setup.
-  - the CNI must be installed before this chart, and `global.cniEnabled: true` must be set.
+To finally deploy linkerd, run
```
Comment (ubergesundheit) [inline README.md:107]: The chart includes an own PSP, so for the bare app to be installed, the `restricted` psp does not need to be edited, so IMO this `finally` could be a little confusing

**Comment by ubergesundheit** [comment]
Tested to be working on KVM 14.1.2 (Kubernetes version 1.19.11) with a test app and linkerd-viz extension

**Pull request merged**